dev-docs/batch-system.rst

```diff
@@ -0,0 +1,784 @@
+   downloads any input files the job needs from cloud storage, creates
+   a container for the job to execute in, executes the code inside the
+   container, uploads any logs and output files that have been
+   generated, and then notifies the Batch driver that the job has
+   completed.
+7. Once all jobs have completed, the batch is set to complete in the
+   database. Any callbacks that have been specified on batch
+   completion are called.
+8. Meanwhile, the user can find the status of their batch through the
+   UI or using a Python client library to get the batch status, cancel
+   the batch, list the jobs in the batch and their statuses, and wait
+   for the batch or an individual job to complete. The implementation
+   of the wait operation is by continuously polling the Batch Front
+   End until the batch state is "complete".
+
+
+~~~~~~~~~~
+Data Model
+~~~~~~~~~~
+
+The core concepts in the Batch data model are billing projects,
+batches, jobs, updates, attempts, and resources.
+
+A **billing project** is a mechanism for imposing cost control and
+enabling the ability to share information about batches and jobs
+across users. Each billing project has a list of authorized users and
+a billing limit. Any users in the billing project can view information
+about batches created in that billing project. Developers can
+add/delete users in a billing project and modify billing limits. Right
+now, these operations are manually done after a Batch user submits a
+formal request to the Hail team. Note that the Hail billing project is
+different than a GCP billing project.
+
+A **batch** is a set of **jobs**. Each batch is associated with a
+single billing project. A batch also consists of a set of
+**updates**. Each update contains a distinct set of jobs. Updates are
+distinct submissions of jobs to an existing batch in the system. They
+are used as a way to add jobs to a batch. A batch is always created
+with 0 updates and 0 total jobs. To add jobs to a batch, an update
+must be created with an additional API call and the number of jobs in
+the update must be known at the time of the API call. The reason for
+this is because an update reserves a block of job IDs in order to
+allow multiple updates to a batch to be submitted simultaneously
+without the need for locking as well as for jobs within the update to
+be able to reference each other before the actual job IDs are
+known. Once all of the jobs for a given batch update have been
+submitted, the update must be committed in order for the jobs to be
+visible in the UI and processed by the batch driver.
+
+A job can have **attempts**. An attempt is an individual execution
+attempt of a job running on a worker VM. There can be multiple
+attempts if a job is preempted. If a job is cancelled before it has a
+chance to run, it will have zero attempts. An attempt has the
+**instance** name that it ran on, the start time, and the end
+time. The end_time must always be greater than the start_time. All
```

Comment:
```suggestion
time. The end time must always be greater than the start time. All
```
Consistency with previous sentence. Prefer no underscore since this is narrative instead of code.